app/authorization/server/functions/hasPermission.js

```diff
@@ -1,31 +1,95 @@
-import { Permissions, Roles } from '../../../models/server/raw';
-
-async function atLeastOne(userId, permissions = [], scope) {
-	for (let i = 0, total = permissions.length; i < total; i++) {
-		const permissionId = permissions[i];
-
-		// eslint-disable-next-line no-await-in-loop
-		const permission = await Permissions.findOne({ _id: permissionId });
-		// eslint-disable-next-line no-await-in-loop
-		const found = await Roles.isUserInRoles(userId, permission.roles, scope);
-		if (found) {
-			return true;
+import mem from 'mem';
+
+import { Permissions, Roles, Users, Subscriptions } from '../../../models/server/raw';
+
+
+const getRole = mem((role) => Roles.findOne({ _id: role }));
+
+const rolesHasPermission = mem(async (permission, roles) => {
+	const result = await Permissions.findOne({ _id: permission, roles: { $in: roles } });
+	return !!result;
+});
+
+const subscriptionHasPermission = mem(async (uid, permission, roles, rid) => {
+	if (rid == null) {
+		return;
+	}
+
+	const query = {
+		'u._id': uid,
+		rid,
+		roles: { $in: roles },
+	};
+
+	return !!await Subscriptions.findOne(query, { fields: { roles: 1 } });
+}, { maxAge: 5000 });
+
+export const clear = () => {
+	mem.clear(getRole);
+	mem.clear(rolesHasPermission);
+	mem.clear(subscriptionHasPermission);
+};
+
+const groupRolesByScope = async (sortedUserRoles) => (await Promise.all(sortedUserRoles.map(getRole))).reduce((roles, role) => {
+	roles[role.scope || 'Users'] = roles[role.scope || 'Users'] ? roles[role.scope || 'Users'].concat([role._id]) : [].concat([role._id]);
+	return roles;
+}, {});
+
+async function atLeastOne(uid, permissions = [], scope) {
+	const { roles: userRoles } = await Users.findOne({ _id: uid });
+
+	const sortedUserRoles = userRoles.sort((a, b) => a.localeCompare(b));
+
+	const roles = await groupRolesByScope(sortedUserRoles);
+	const keys = Object.keys(roles);
+	for (let index = 0; index < keys.length; index++) {
+		const key = keys[index];
+		switch (key) {
+			case 'Users':
+				for (const permission of permissions) {
+					if (await rolesHasPermission(permission, roles[key])) { // eslint-disable-line
+						return true;
+					}
+				}
+				break;
+			case 'Subscriptions':
+				const found = await subscriptionHasPermission(uid, roles[key], scope);// eslint-disable-line
+				if (found) {
+					return true;
+				}
+				break;
 		}
 	}
 
 	return false;
 }
 
-async function all(userId, permissions = [], scope) {
-	for (let i = 0, total = permissions.length; i < total; i++) {
-		const permissionId = permissions[i];
+async function all(uid, permissions = [], scope) {
+	const { roles: userRoles } = await Users.findOne({ _id: uid });
+
+	const sortedUserRoles = userRoles.sort((a, b) => a.localeCompare(b));
```

Comment:
Why sort here?

Reply:
we are using the combination of roles to caching a result... to avoid an `[user, admin]` and a `[admin, user]` I'm sorting